app/javascript/javascript/mapsUtils.js

```diff
@@ -4,3 +4,59 @@ export const lengthToPixels = (length, zoom) => {
 };
 
 export const pixelsToLength = (pixels, zoom) => pixels * Math.pow(2, -zoom);
+
+export const clearMap = () => {
+  if (process.env.NODE_ENV !== "test") {
+    window.__map.clearMarkers();
+    window.__map.clearInfoWindows();
+    window.__map.clearPolygons();
+  }
+};
+
+if (process.env.NODE_ENV !== "test") {
+  // Polygons
+  google.maps.Map.prototype.polylines = new Array();
+  google.maps.Map.prototype.rectangles = new Array();
+
+  google.maps.Map.prototype.clearPolygons = function() {
+    this.polylines.forEach(polyline => polyline.setMap(null));
+    this.polylines = new Array();
+    this.rectangles.forEach(rectangle => rectangle.setMap(null));
+    this.rectangles = new Array();
+  };
+
+  // InfoWindows
+  google.maps.Map.prototype.infoWindows = new Array();
+
+  google.maps.Map.prototype.clearInfoWindows = function() {
+    this.infoWindows.forEach(infoWindow => infoWindow.close());
+    this.infoWindows = new Array();
+  };
+
+  google.maps.InfoWindow.prototype._open =
+    google.maps.InfoWindow.prototype.open;
+
+  google.maps.InfoWindow.prototype.open = function(map, marker) {
+    if (map) {
+      map.infoWindows.push(this);
+    }
+    this._open(map, marker);
+  };
+
+  // Markers
+  google.maps.Map.prototype.markers = new Array();
+
+  google.maps.Map.prototype.clearMarkers = function() {
+    this.markers.forEach(marker => marker.setMap(null));
+    this.markers = new Array();
+  };
+
+  google.maps.Marker.prototype._setMap = google.maps.Marker.prototype.setMap;
+
+  google.maps.Marker.prototype.setMap = function(map) {
+    if (map) {
+      map.markers.push(this);
+    }
+    this._setMap(map);
+  };
+}
```

Comment:
I'm not a big fan of modifying prototypes. I think it's great if this is a step towards centralising the logic. In the end I think we should wrap Google Maps instead of modifying it.

If we had all of the calls interacting with google maps in one place then we could keep track of things in there encapsulated.

Reply:
I agree modifying is not ideal (although I found this approach as a recommended solution on StackOverflow.) On the other hand, using a wrap still, leave a possibility that someone unfamiliar with a project will use `marker.setMap(mapObj)` directly, and then have a bug with this marker not being cleared from the map.

Reply:
And in this case, these bugs are hard to notice (like when one marker is on top of it's duplicate or if the bug only happens when switching filters in a specific order).
What I mean is that a solution where it's impossible to introduce a bug is better than a solution that relies on developers following good practices 😉 
Anyway, next time there is an opportunity to refactor that code we can do some brainstorming :)